src/bvals/comms/coalesced_buffers.cpp

```diff
@@ -0,0 +1,482 @@
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#include <cstdio>
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "basic_types.hpp"
+#include "bvals/comms/bnd_id.hpp"
+#include "bvals/comms/bvals_utils.hpp"
+#include "bvals/comms/coalesced_buffers.hpp"
+#include "bvals/neighbor_block.hpp"
+#include "coordinates/coordinates.hpp"
+#include "interface/mesh_data.hpp"
+#include "interface/variable.hpp"
+#include "mesh/mesh.hpp"
+#include "mesh/meshblock.hpp"
+#include "utils/communication_buffer.hpp"
+
+namespace parthenon {
+
+//----------------------------------------------------------------------------------------
+void CoalescedBuffer::AllocateCoalescedBuffer() {
+  int send_rank = sender ? Globals::my_rank : other_rank;
+  int recv_rank = sender ? other_rank : Globals::my_rank;
+  coalesced_comm_buffer = CommBuffer<buf_t>(
+      2 * partition, send_rank, recv_rank, comm_,
+      [](int size) { return buf_t("Combined Buffer", 2 * size); }, true);
+
+  sparse_status_buffer =
+      CommBuffer<std::vector<int>>(2 * partition + 1, send_rank, recv_rank, comm_);
+  sparse_status_buffer.ConstructBuffer(current_size + 1);
+}
+
+//----------------------------------------------------------------------------------------
+ParArray1D<BndId> &CoalescedBuffer::GetBndIdsOnDevice(const std::set<Uid_t> &vars,
+                                                      int *pcomb_size) {
+  const auto &var_set = vars.size() == 0 ? all_vars : vars;
+  auto &bnd_ids_device = bnd_ids_device_map[var_set];
+  auto &bnd_ids_host = bnd_ids_host_map[var_set];
+
+  int nbnd_id{0};
+  int comb_size{0};
+  for (auto uid : var_set) {
+    // Skip this variable if it is not communicated in this BoundaryType
+    if (coalesced_info_buf.count(uid) == 0) continue;
+    nbnd_id += coalesced_info_buf.at(uid).size();
+    for (auto &[bnd_id, pvbbuf] : coalesced_info_buf.at(uid)) {
+      auto buf_state = pvbbuf->GetState();
+      if ((buf_state == BufferState::sending) || (buf_state == BufferState::received))
+        comb_size += bnd_id.size();
+    }
+  }
+  if (pcomb_size != nullptr) *pcomb_size = comb_size;
+
+  bool updated = false;
+  if (comb_size > coalesced_comm_buffer.buffer().size()) {
+    PARTHENON_REQUIRE(
+        sender, "Something bad is going on if we are doing this on a receiving buffer.");
+    coalesced_comm_buffer.ConstructBuffer("combined send buffer", 2 * comb_size);
+    updated = true;
+  }
+
+  if (nbnd_id != bnd_ids_device.size()) {
+    bnd_ids_device = ParArray1D<BndId>("bnd_id", nbnd_id);
+    bnd_ids_host = Kokkos::create_mirror_view(bnd_ids_device);
```

Comment:
Regarding the view of view piece, this may need to be
 `ParArrayRaw1D<BndId>(ViewOfViewAlloc("bnd_id"), nbnd_id);`
and
`create_view_of_view_mirror(bnd_ids_device)`

Reply:
Thank you for pointing this out, it was definitely the cause of at least one bug in Riot that I was chasing.